Update LIBFFI for z/OS for varargs

r30shah opened the pull request:
In case of variadic arguments in XPLINK, all arguments within the variadic part are passed into GPRs if available or argument list even if arguments are of double type. z/OS port of Libffi was missing handling of such variadic arguments, causing us to pass double or complex type parameters into FPRs. This commit adds new field in ffi_cif struct providing number of fixed args in parameters and use it to ensure double or complex type parameters within variadic part is set up correctly.

pshipton commented:
If any of this was taken from the libffi project, we'll need to update various places that refer to a libffi version and copyright.
i.e. at least
https://github.com/eclipse-openj9/openj9/blob/master/LICENSE
https://github.com/eclipse-openj9/openj9/blob/master/longabout.html
https://github.com/eclipse-openj9/openj9/blob/master/runtime/libffi/LICENSE
https://github.com/eclipse-openj9/openj9/blob/master/runtime/libffi/about.html

r30shah commented:
> If any of this was taken from the libffi project, we'll need to update various places that refer to a libffi version and copyright.

Thanks. I haven't looked into recent libffi code (It does not have z/OS port). Changes made in this PR, adds a field, provision of which was already there if needed which is already done on some other platform in the version of libffi we have. Will double check if we need to update those spaces or not.

r30shah commented:
@keithc-ca  Can I get your review on this change?

r30shah commented:
IBM SDK 8 test build for z/OS : `build_info.php?build_id=99998`

r30shah commented on runtime/libffi/z/ffi.c:
```diff
@@ -267,31 +267,36 @@ ffi_prep_args (unsigned char *stack, extended_cif *ecif)
 
 /*======================== End of Routine ============================*/
- 
+ffi_status
```
In `ffi.c` file, this is the only change - I have not yet changed anything functionally for 31-bit z/OS (Given that we do not have support for struct parameters in 31-bit z/OS) - Though the variadic argument change for double needs to be changed here - I need to check with ABI document w.r.t 31-bit z/OS before making any functional change.

Rest of the changes in the files are removal of white space at the end of line.

r30shah commented on runtime/libffi/z/ffi.c:
```diff
@@ -114,27 +114,27 @@ ffi_check_struct_type (ffi_type *arg)
        	return FFI_TYPE_FLOAT;
       else
-       	return FFI_TYPE_UINT32; 
+       	return FFI_TYPE_UINT32;
```
Hmm, it complains about this and line 127
```
09:51:27  runtime/libffi/z/ffi.c:122: space before tab in indent.
09:51:27  +       	return FFI_TYPE_UINT32;
09:51:27  runtime/libffi/z/ffi.c:127: space before tab in indent.
09:51:27  +       	return FFI_TYPE_UINT64;
```

I am not sure what change I can make here - I simply removed whitespace at the end of file.

r30shah commented on runtime/libffi/z/ffi64.c:
```diff
@@ -821,12 +833,13 @@ ffi_check_struct_for_complex(ffi_type *arg_type)
 
 ffi_status
-ffi_prep_cif_machdep(ffi_cif *cif)
+ffi_prep_cif_machdep_core(ffi_cif *cif)
```
Moved the code in  https://github.com/eclipse-openj9/openj9/pull/22396/commits/9d4c27b3ffb0a7de0b74000a5319985ad10c1baa

keithc-ca replied:
The comment leader for this function should be updated to match its new name.

r30shah commented on runtime/libffi/z/ffi64.c:
```diff
@@ -945,11 +963,21 @@ ffi_prep_cif_machdep(ffi_cif *cif)
 	  /* If the ffi_type is STRUCT with two floating point parameters of the same type
-	     they will be passed in available FPRs. Logic to calculate bytes for the struct
-	     should be same as normal struct.
+	     they will be passed in available FPRs except argument is part of variadic argument list.
```
Changed in https://github.com/eclipse-openj9/openj9/pull/22396/commits/9d4c27b3ffb0a7de0b74000a5319985ad10c1baa

r30shah commented on runtime/libffi/z/ffitarget.h:
```diff
@@ -66,6 +66,7 @@ typedef enum ffi_abi {
 #endif
 #define FFI_NATIVE_RAW_API 0
-
+#define FFI_EXTRA_CIF_FIELDS unsigned z_nfixedargs
```
I went with how other unsigned int parameters of the struct through out the libffi code is set [1].

[1]. https://github.com/eclipse-openj9/openj9/blob/bea400954cbce38c648db3f4177d94daebf648ff/runtime/libffi/preconf/mz64/ffi.h#L231-L241

keithc-ca replied:
Those ought to be corrected as well: `z_nfixedargs`, `nargs` and the second and third parameters to `ffi_prep_cif_machdep_var()` should all be the same type.

keithc-ca commented on runtime/libffi/z/ffi64.c:
```diff
@@ -811,7 +811,6 @@ ffi_check_struct_for_complex(ffi_type *arg_type)
   }
   return arg_type->type;
 }
-
```
Please don't delete separating blank lines such as this. There should also be an "end" marker like other functions:
```c

/*======================== End of Routine ============================*/

```

keithc-ca commented on runtime/libffi/z/ffi64.c:
```diff
@@ -1012,7 +1027,19 @@ ffi_prep_cif_machdep(ffi_cif *cif)
 }
 
 /*======================== End of Routine ============================*/
+ffi_status
+ffi_prep_cif_machdep_var(ffi_cif *cif, unsigned int nfixedargs, unsigned int ntotalargs)
+{
+  cif->z_nfixedargs = nfixedargs;
+  return ffi_prep_cif_machdep_core(cif);
+}
 
+ffi_status
+ffi_prep_cif_machdep(ffi_cif *cif)
+{
+  cif->z_nfixedargs = cif->nargs;
+  return ffi_prep_cif_machdep_core(cif);
+}
```
Both of these functions should have the comments before and after.

r30shah commented on runtime/libffi/z/ffi64.c:
```diff
@@ -1013,6 +1031,38 @@ ffi_prep_cif_machdep(ffi_cif *cif)
+/* Function - Perform machine dependent CIF processing for varargs.   */
+/*                                                                    */
+/*====================================================================*/
+ffi_status
```
Fixed in https://github.com/eclipse-openj9/openj9/pull/22396/commits/8af5f1e6c40f76b42767dc14195edab38431fc90

r30shah commented on runtime/libffi/z/ffi.c:
```diff
@@ -452,20 +451,37 @@ ffi_prep_cif_machdep(ffi_cif *cif)
+ffi_prep_cif_machdep_var(ffi_cif *cif, unsigned int nfixedargs, unsigned int ntotalargs)
+{
+  cif->z_nfixedargs = nfixedargs;
+  return ffi_prep_cif_machdep(cif);
```
Changed in https://github.com/eclipse-openj9/openj9/pull/22396/commits/c75c2f942ae4264d31d12968fc60a3164052288d

r30shah commented on runtime/libffi/z/ffitarget.h:
```diff
@@ -66,6 +66,7 @@ typedef enum ffi_abi {
+#define FFI_EXTRA_CIF_FIELDS unsigned int z_nfixedargs
+#define FFI_TARGET_SPECIFIC_VARIADIC 1
 #endif
 
```
Cleaned up in https://github.com/eclipse-openj9/openj9/pull/22396/commits/17a6f50216a0ed2304a1ec8b01047a4152401c4d

r30shah commented:
@keithc-ca  I have tried to address most of comments, Let me know if changes are good or not, I am planning to squash the commits and do one more internal sanity runs after making all changes.

keithc-ca commented on runtime/libffi/z/ffi.c:
```diff
@@ -267,33 +267,33 @@ ffi_prep_args (unsigned char *stack, extended_cif *ecif)
+
 ffi_status
-ffi_prep_cif_machdep(ffi_cif *cif)
+ffi_prep_cif_machdep_core(ffi_cif *cif)
```
This should be static, like the 64-bit version.

r30shah replied:
Fixed in https://github.com/eclipse-openj9/openj9/pull/22396/commits/523147ec00bb79f9113161a74fb9254769aa98f5

keithc-ca commented on runtime/libffi/z/ffi.c:
```diff
@@ -452,20 +452,54 @@ ffi_prep_cif_machdep(ffi_cif *cif)
+/*                                                                    */
+/* Name     - ffi_prep_cif_machdep.                                   */
+/*                                                                    */
+/* Function - Perform machine dependent CIF processing                */
```
Missing period.

r30shah replied:
Fixed in https://github.com/eclipse-openj9/openj9/pull/22396/commits/523147ec00bb79f9113161a74fb9254769aa98f5

keithc-ca commented on runtime/libffi/z/ffi.c:
```diff
@@ -267,33 +267,33 @@ ffi_prep_args (unsigned char *stack, extended_cif *ecif)
+/* Name     - ffi_prep_cif_machdep_core.                              */
 /*                                                                    */
-/* Function - Perform machine dependent CIF processing.               */
+/* Function - Core implementation of machine dependent CIF processing.*/
```
Please adjust this so there's at least one space before `*/` (perhaps by splitting the line).

r30shah replied:
Fixed in https://github.com/eclipse-openj9/openj9/pull/22396/commits/523147ec00bb79f9113161a74fb9254769aa98f5

keithc-ca commented on runtime/libffi/z/ffi64.c:
```diff
@@ -812,21 +812,24 @@ ffi_check_struct_for_complex(ffi_type *arg_type)
+/* Name     - ffi_prep_cif_machdep_core.                              */
 /*                                                                    */
-/* Function - Perform machine dependent CIF processing.               */
+/* Function - Core implementation of machine dependent CIF processing.*/
```
Please adjust this so there's at least one space before */ (perhaps by splitting the line).

r30shah replied:
Fixed in https://github.com/eclipse-openj9/openj9/pull/22396/commits/523147ec00bb79f9113161a74fb9254769aa98f5

r30shah commented on runtime/libffi/z/ffi.c:
```diff
@@ -267,33 +267,33 @@ ffi_prep_args (unsigned char *stack, extended_cif *ecif)
+/*                                                                     */
+/* Name     - ffi_prep_cif_machdep_core.                               */
+/*                                                                     */
+/* Function - Core implementation of machine dependent CIF processing. */
```
Changed in https://github.com/eclipse-openj9/openj9/pull/22396/commits/4d1a32ede89811dfe5cf22fbd5c78226f4ac9bdb

keithc-ca reviewed (changes requested):
Please squash. When updating the commit message, be sure to follow the guidelines documented [here](https://github.com/eclipse-openj9/openj9/blob/master/CONTRIBUTING.md#commit-guidelines). In particular, lines in the body of that message should be no more than 72 characters long.

r30shah commented:
@keithc-ca  this is ready for review. I have squashed all previous commits to https://github.com/eclipse-openj9/openj9/pull/22396/commits/6e32e79e72b73d055c38cc33f88267367248bece and added https://github.com/eclipse-openj9/openj9/pull/22396/commits/de5190ba4c47ae7c026d80a562fe7d0bde587cdb to handle double type vararg in assembly code.

r30shah commented:
@joransiu  Can I get your review as well with this change?

r30shah commented:
> Was the testing mentioned in https://github.com/eclipse-openj9/openj9/pull/22396#issuecomment-3198126021 done with this in its current form or something earlier?

There was a minor code improvement I made this morning, so I tested changes early this morning with latest changes as well.

keithc-ca commented on runtime/libffi/z/sysvz64.s:
```diff
@@ -387,6 +387,12 @@ FLTSTR   DS 0H
+         L  15,32(,15)        cif->z_nfixedargs
+         SLL 15,3             offset of first vararg type
+         CR  10,15            argOffset >= z_nfixedargs << 3
+         BRNL SI64            Double from vararg, use GPR
```
This has the best information I could find: https://www.ibm.com/docs/en/zos/3.1.0?topic=SSLTBW_3.1.0/com.ibm.zos.v3r1.ceev100/cee1v292.html. The way I read it, I think we need to handle `float` similar to how this handles `double`.

keithc-ca replied:
The part quoted above is from a section qualified with "calls to unprototyped functions, where the caller cannot know whether the called function contains a variable (vararg) portion". I don't think that applies here because we're treating the call description as a prototype of the target function. On the other hand, in cases where there are no fixed parameters it would probably be safer to operate as if we have no prototype.

r30shah replied:
> Do we need similar logic to handle vararg float parameters?

I was trying to find the relevant documentation that explicitly mentions promotion of floats to double if they are from vararg part - can not find the xlc/openXL document for that - but I implemented the unit test and tested that with both OpenXL compiler and xlc compiler on z/OS. 

```
#include <stdarg.h>
#include <stdlib.h>
#include <stdio.h>


float varargsWrong(int num, ...) {
    printf("In Wrong\n");
    float a = 1.0F;
    va_list a_list;
    va_start(a_list, num);
    for (int i = 0; i < num; i++) {
	float x = va_arg(a_list, float);
	printf("Passed float %f\n",x);
	a = a* x;
    }
    va_end(a_list);
    return a;
}

double varargsRight(int num, ...) {
    printf("In Right\n");
    double a = 1.0F;
    va_list a_list;
    va_start(a_list, num);
    for (int i = 0; i < num; i++) {
	double x = va_arg(a_list, double);
	printf("Passed float %lf\n",x);
	a = a* x;
    }
    va_end(a_list);
    return a;
}



int main() {
	printf("In main\n");
	varargsWrong(3, 1.23456F, 4.5678F, 98.6384F);
	varargsRight(3, 1.23456F, 4.5678F, 98.6384F);
	return 0;
}
```

Output : 
```
In main
In Wrong
Passed float 0.000000
Passed float 36893488147419103232.000000
Passed float -0.000000
In Right
Passed float 1.234560
Passed float 4.567800
Passed float 98.638397
```

So when I try to read the arguments as float, I got incorrect values. xlc did not spit any warning, but OpenXL was explicit in warnings suggesting reading va_arg as float is undefined. 

```
libVarArgs.c:12:27: warning: second argument to 'va_arg' is of promotable type 'float'; this va_arg has undefined behavior because arguments will be promoted to 'double' [-Wvarargs]
   12 |         float x = va_arg(a_list, float);
      |                                  ^~~~~
/usr/include/stdarg.h:106:57: note: expanded from macro 'va_arg'
  106 |        #define va_arg(ap, type)    __builtin_va_arg(ap, type)
      |                                                         ^~~~
1 warning generated.
```

I do not expect a ffi_type_float set for the argument that is part of var_arg.

r30shah replied:
> A successful symbol lookup doesn't guarantee that the signature is correct, so in some senses, we're only making unprototyped calls.

Ok, so thinking more about it, a method handle for function call with vararg can be generated and used to call a regular function which does not take vararg. So in that case, we do want to set-up all GPRs/ArgumentList and FPRs in speculation of unprototyped calls, change will be lil more complicated, and would need some more complex and careful changes in HLA.  

Do you think with JNI it is possible to call unprototyped function? If it is not possible, then would it be ok to follow up that change as a separate PR as I think that change would need lot more testing.

r30shah replied:
> I think making change in the HLA to add same checks as we are doing in double would be simpler - With explicit promotion - I do fear reusing of cached cif object later on for normal call may cause issue (Which I did observe when I was changing the type DOUBLE to SI64 in prep_cif. I will add this change in this PR.

Regarding to handling `FFI_TYPE_FLOAT` in vararg, only concern I have is that it is going to make the common case where we are passing float to normal function slower (As it would need us to load the ecif->cif->z_nfixedargs on every float type and compare against argID - which if code is written correctly should never happen).
